gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,58 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar, Union
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData, OutputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for models that encapsulate their data.
+        """
+        return -self.maximum_a_posteriori_objective()
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingLossMixin:
+    def training_loss(self, data: RegressionData) -> tf.Tensor:
+        """
+        Training loss for models that do not encapsulate the data.
+        
+        :param data: the data to be used for computing the model objective.
+        """
+        return -self.maximum_a_posteriori_objective(data)
+
+    def training_loss_closure(
+        self, data: Union[Data, collections.abc.Iterator], jit=True
+    ) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                training_loss
+            )  # TODO need to add correct input_signature here to allow for differently sized minibatches
+
+        def closure():
+            batch = next(data) if isinstance(data, collections.abc.Iterator) else data
```

Comment:
see my [comment](https://github.com/GPflow/GPflow/pull/1362/files#r397557765) on eric's PR about this

Reply:
For optimization use-cases, the dataset iterators only really make sense when being `.repeat()`ed, so they never run out. If someone doesn't do that and next() raises, it's up to the caller to handle it - I don't think we need to do any more here.

Reply:
note [`repeat`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#repeat) has an argument `count` to specify the number of repeats

Reply:
I still think it's best to simply not handle the StopIteration and let the caller figure out what to do with it. Seems like the most pythonic way to me, anyways